### PR TITLE
IGNITE-16604: Fix GeoSpatial indexes for client nodes

### PR DESCRIPTION
**Pull request by timoninmaxim**
Client nodes incorrectly handle geo-spatial indexes, and fail on trying to create index. This patch provides a mock for geo-spatial indexes on client nodes by analogue with ClientInlineIndex for sorted indexes. It enables creation of indexes on Ignite client nodes.

### CODE REVIEW

File: modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndex.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        if (clazz == null)
+            return null;
+
+        if (clazz.isAssignableFrom(getClass()))
+            return clazz.cast(this);
+
+        throw new IllegalArgumentException(
+            String.format("Cannot unwrap [%s] to [%s]", getClass().getName(), clazz.getName())
+        );
+    }
```
Comment (anton-vinogradov) [inline modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndex.java:105]: Looks like a dupliccaton across all Index implementations.
Could this be implemented as a default method?

File: modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndexDefinition.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    private final LinkedHashMap<String, IndexKeyDefinition> keyDefs;
+
+    /** */
+    private final IndexName idxName;
+
+    /** */
+    private final GridH2Table tbl;
+
+    /** */
+    private final List<IndexColumn> cols;
```
Comment (anton-vinogradov) [inline modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndexDefinition.java:41]: We must be pretty sure to keep only required data. 
Please recheck all these fields are used at client node.

File: modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndex.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    private final UUID id = UUID.randomUUID();
+
+    /** */
+    private final String name;
+
+    /** */
+    private final GridH2Table tbl;
+
+    /** */
+    private final List<IndexColumn> cols;
```
Comment (anton-vinogradov) [inline modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndex.java:49]: Could we keep GeoSpatialClientIndexDefinition instead?

File: modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java [hunk 4]
```diff
@@ -76,6 +80,28 @@
     /** Segmented index flag. */
     private final boolean segmented;
 
+    /** Node that creates a Geo-Spatial index. */
+    @Parameterized.Parameter
+    public String crtNode;
+
+    /** Node that stars a query over a Geo-Spatial index. */
+    @Parameterized.Parameter(1)
+    public String qryNode;
```
Comment (anton-vinogradov) [inline modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java:89]: Please simplify to boolean couple and use them to select client node or server.

File: modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java [hunk 5]
```diff
@@ -86,15 +112,27 @@ protected H2IndexingAbstractGeoSelfTest(boolean segmented) {
     }
 
     /** {@inheritDoc} */
-    @Override protected int gridCount() {
-        return 3;
+    @Override public void beforeTest() throws Exception {
+        startGrids(3);
```
Comment (anton-vinogradov) [inline modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java:116]: assign `srv` node here as well as `cln` ant use it further

File: modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java [hunk 6]
```diff
@@ -122,11 +159,12 @@ protected H2IndexingAbstractGeoSelfTest(boolean segmented) {
      * @return Cache.
      * @throws Exception If failed.
      */
-    @SuppressWarnings("unchecked")
     protected <K, V> IgniteCache<K, V> createCache(String name, boolean partitioned, Class<?> keyCls, Class<?> valCls,
         boolean dynamicIdx) throws Exception {
         CacheConfiguration<K, V> ccfg = cacheConfig(name, partitioned, keyCls, valCls);
 
+        IgniteEx crtIgn = node(crtNode);
```
Comment (anton-vinogradov) [inline modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java:166]: use ternar operation here to chose between `cln` and `srv`

File: modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java [hunk 7]
```diff
@@ -629,11 +663,11 @@ public Enemy(int campId, String name) {
     protected static class EnemyCamp implements Serializable {
         /** */
         @QuerySqlField(index = true)
-        Geometry coords;
+        final Geometry coords;
```
Comment (anton-vinogradov) [inline modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java:666]: protected?

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java [hunk 8]
```diff
@@ -182,7 +182,7 @@
 
     /** Index manager. */
     @GridToStringExclude
-    private IndexProcessor idxMgr;
+    private final IndexProcessor idxProc;
```
Comment (anton-vinogradov) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java:185]: Good chance to make all fields final where possible

**Review by anton-vinogradov — CHANGES_REQUESTED** [review]

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/index/client/ClientInlineIndex.java [hunk 9]
```diff
@@ -148,7 +148,7 @@ public ClientInlineIndex(String idxName, int inlineSize) {
     /**
      * @return Exception about unsupported operation.
      */
-    private static IgniteException unsupported() {
+    public static IgniteException unsupported() {
```
[the remark this answers is not in the document]
Comment (timoninmaxim) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/index/client/ClientInlineIndex.java:151]: I'm not sure that it's OK to move this method to U class. This method is expected to be used only by client nodes. IMHO, the better option is an additional layer like `AbstractClientIndex` with this method.

File: modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java [hunk 10]
```diff
@@ -76,6 +80,28 @@
...
+    private IgniteEx cln;
+
+    /** */
+    @Parameterized.Parameters(name = "crtNode={0}, qryNode={1}")
+    public static List<Object[]> parameters() {
+        return F.asList(
+            new Object[] { "CRD", "CRD" },
+            new Object[] { "CRD", "CLN" },
+            new Object[] { "CLN", "CRD" },
+            new Object[] { "CLN", "CLN" }
```
[the remark this answers is not in the document]
Comment (timoninmaxim) [reply modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java:101]: There is even better solution - use cartesian product of array `{ "CLN", "CRD" }` on itself.

File: modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndexDefinition.java (hunk 2 above)
Comment (timoninmaxim) [reply modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndexDefinition.java:41]: Don't understand your concern, it looks like all those fields are required.

File: modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java (hunk 7 above)
Comment (timoninmaxim) [reply modules/geospatial/src/test/java/org/apache/ignite/internal/processors/query/h2/H2IndexingAbstractGeoSelfTest.java:666]: not, it used in `checkDistributedQuery()` by direct reference.

File: modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialUtils.java [hunk 11]
```diff
@@ -46,27 +46,53 @@
...
     /** */
-    public static GridH2SpatialIndex createIndex(GridH2Table tbl, String idxName, List<IndexColumn> cols) {
+    public static GridH2IndexBase createIndex(GridH2Table tbl, String idxName, List<IndexColumn> cols) {
         try {
             IndexName name = new IndexName(tbl.cacheName(), tbl.getSchema().getName(), tbl.getName(), idxName);
 
             LinkedHashMap<String, IndexKeyDefinition> keyDefs = new QueryIndexKeyDefinitionProvider(tbl, cols).keyDefinitions();
 
-            List<InlineIndexKeyType> idxKeyTypes = InlineIndexKeyTypeRegistry.types(keyDefs.values(), DUMMY_SETTINGS);
+            if (tbl.cacheInfo().affinityNode())
```
[the remark this answers is not in the document]
Comment (timoninmaxim) [reply modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialUtils.java:55]: No, node filters are checked before the index creation procedure.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/index/client/ClientInlineIndex.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (anton-vinogradov) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/index/client/ClientInlineIndex.java]: Any ancestor also is good choice

File: modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndexDefinition.java (hunk 2 above)
Comment (anton-vinogradov) [reply modules/geospatial/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GeoSpatialClientIndexDefinition.java:41]: Just a recheck proposal.

**Review by anton-vinogradov — APPROVED** [review]

**Pull request merged**